### PR TITLE
les: move client pool to les/vflux/server

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR refactors the client peer logic for LES servers similarly to how server peer logic was recently refactored on the client side; now all complex peer logic is separated from the protocol implementation and implemented in the `vflux` subpackages. The NodeStateMachine is only used in these packages and all references have been removed from package `les` where a very simple `serverPeerSet` and `clientPeerSet` are representing the currently active peers. These changes also prepare the codebase in order to make the next steps (specifically the capacity update logic of `les/5`) easier to deploy.
A fuzzer for `vfs.ClientPool` is also implemented in this PR which covers most of the `vfs` package. This fuzzer does much more meaningful testing of the peer logic than fuzzing the `NodeStateMachine` alone.

Note: `vfs.BalanceTracker.BalanceOperation` allows atomic balance operations in the sense that the in-memory peer structure will not change between operations. The database storage of the changed balance is not atomic yet, this will be implemented in the next PR that adds the token sale mechanism. This is where this feature will really make sense because the ether balance and token balance needs to be updated atomically.

### CODE REVIEW

File: les/vflux/server/balance_tracker.go [hunk 1]
```diff
@@ -40,7 +41,7 @@ type BalanceTrackerSetup struct {
 	PriorityFlag, UpdateFlag nodestate.Flags
 	BalanceField             nodestate.Field
 	// external connections
-	connAddressField, capacityField nodestate.Field
```
Comment (rjl493456442) [inline les/vflux/server/balance_tracker.go:43]: I think my first suggestion to this big PR is: now we can lowercase these fields and flags!

Comment (rjl493456442) [reply les/vflux/server/balance_tracker.go:43]: Or is there some elegant approaches we can get rid of the `Connect` function? Now all the components are in the same package.

Comment (zsfelfoldi) [reply les/vflux/server/balance_tracker.go:43]: We can actually lowercase most of the package now :) I kept the connect function though because I use different test client types and corresponding test fields for priorityPool and balanceTracker tests which are wired together statically. Of course we can change it if you have a more elegant idea.

**Comment by samlavery** [comment]
My only actual question is who to git blame, and do you have any evidence of intent to supress the non-protected rights of tokens? Or does the committer have a known history of making type-o’a and other grammar type mistakes?  I make simple mistakes all the time, but I didn’t mean to do it.  I would need much more than some git commits to be able to infer anything. It would take a long time to unwind all of this code I haven’t seen in years. 

Thanks,
Sam

**Comment by samlavery** [comment]
So I've never liked Kademlia, for known historical reasons, but I don't
even know what you are using for your network formation these days? What do
changes to code that evaluates context have to do with code that manages
the nodes you connect with?  These should be in two separate pull requests,
right?  It makes it easier to review, which is kind of the point of
the whole open source thing, I guess.  Are we going to need to make
NATIONAL CODE REVIEW JURY SERVICE A THING NOW?

This is a common way for bad things to make their way into reasonable
code.  It has happened before, and it probably will happen again.

**Comment by samlavery** [comment]
So I'm guessing we missed a class on information theory, compression and
storage costs?
Am I getting warmer?

**Comment by samlavery** [comment]
I know this may offend some people, but the often cantankerous consensus
mechanism of large mutli-platform decentralized systems is one of its
strengths, not always a weakness.  Massive coordinated updates are always
concerning. Emergency patches are fine, sudden updates can be troublesome.


If we all reacted the same way, we'd be predictable, and there's always
more than one way to view a situation. What's true for the group is also
true for the individual. It's simple: Overspecialize, and you breed in
weakness. It's slow death.

File: les/vflux/server/prioritypool.go [hunk 2]
```diff
@@ -206,14 +198,14 @@ func (pp *PriorityPool) RequestCapacity(node *enode.Node, targetCap uint64, bias
...
+		log.Error("requestCapacity called for unknown node", "id", node.ID())
 		return math.MaxInt64, false
 	}
 	var priority int64
 	if targetCap > c.capacity {
-		priority = c.nodePriority.EstimatePriority(targetCap, 0, 0, bias, false)
+		priority = c.nodePriority.estimatePriority(targetCap, 0, 0, bias, false)
 	} else {
-		priority = c.nodePriority.Priority(targetCap)
+		priority = c.nodePriority.priority(targetCap)
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:208]: If the requested capacity is lower than the current one, should we just return "YES"?

File: les/vflux/server/prioritypool.go [hunk 3]
```diff
@@ -224,13 +216,13 @@ func (pp *PriorityPool) RequestCapacity(node *enode.Node, targetCap uint64, bias
 	_, minPriority = pp.enforceLimits()
 	// if capacity update is possible now then minPriority == math.MinInt64
 	// if it is not possible at all then minPriority == math.MaxInt64
-	allowed = priority > minPriority
+	allowed = priority >= minPriority
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:219]: Is it possible to return the update list from the `enforceLimits`? It's weird to save it in the priority queue.

File: les/vflux/server/clientpool.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+		if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+			newCap, _ := newValue.(uint64)
+			c.UpdateCapacity(newCap, node == cp.capReqNode)
+		}
+	})
+	return cp
+}
+
+// AddMetrics adds metrics to the client pool. Should be called before Start().
+func (cp *ClientPool) AddMetrics(totalConnectedGauge metrics.Gauge,
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:166]: Can we define these client pool internal metrics under the package `vflux/server`? We don't need to put all the metrics in a single place.

File: les/vflux/server/clientpool.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+			// (will be activated automatically with minCap when possible)
+			return
+		}
+		if reqCap < cp.minCap {
+			// can't request less than minCap; switching between 0 (inactive state) and minCap is
+			// performed by the server automatically as soon as necessary/possible
+			reqCap = cp.minCap
+		}
+		if reqCap > cp.minCap {
+			if cp.ns.GetState(node).HasNone(btSetup.priorityFlag) && reqCap > cp.minCap {
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:260]: `reqCap > cp.minCap` is checked for two times

File: les/vflux/server/clientpool.go [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+			return
+		}
+		curveBias := bias
+		if requested {
+			// mark the requested node so that the UpdateCapacity callback can signal
+			// whether the update is the direct result of a SetCapacity call on the given node
+			cp.capReqNode = node
+			defer func() {
+				cp.capReqNode = nil
+			}()
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:275]: Can we do the notification in this function(for the capacity updating)? It's unnecessary to rely on an external callback to this...

File: les/vflux/server/clientpool.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+		}
+
+		// estimate maximum available capacity at the current priority level and request
+		// the estimated amount; allow a limited number of retries because individual
+		// balances can change between the estimation and the request
+		for count := 0; count < 20; count++ {
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:281]: I'm not sure the retry is necessary. Serving this function should be quick enough.

File: les/vflux/server/balance.go [hunk 8]
```diff
@@ -78,7 +114,62 @@ type NodeBalance struct {
...
+
+// negValue returns the value of negative balance at a given timestamp.
+func (b balance) negValue(now mclock.AbsTime) uint64 {
+	return b.neg.Value(b.negExp.LogOffset(now))
+}
+
+// add adds the value of a given amount to the balance. The original value and
+// updated value will also be returned if the addition is successful.
+// Returns the error if the given value is too large and the value overflows.
+func (b *balance) addPosValue(now mclock.AbsTime, amount int64, pos bool, force bool) (uint64, uint64, int64, error) {
```
Comment (rjl493456442) [inline les/vflux/server/balance.go:134]: Please rename it to `addValue`. This function is used for positive and negative balance.

**Review by alisscco — APPROVED** [review]

File: les/vflux/server/prioritypool.go [hunk 9]
```diff
@@ -451,31 +458,31 @@ func (pp *PriorityPool) finalizeChanges(commit bool) (updates []capUpdate) {
...
 		if f.oldCap == 0 {
-			pp.ns.SetStateSub(f.node, pp.ActiveFlag, pp.InactiveFlag, 0)
+			pp.ns.SetStateSub(f.node, pp.activeFlag, pp.inactiveFlag, 0)
 		}
 		if f.newCap == 0 {
-			pp.ns.SetStateSub(f.node, pp.InactiveFlag, pp.ActiveFlag, 0)
-			pp.ns.SetFieldSub(f.node, pp.CapacityField, nil)
+			pp.ns.SetStateSub(f.node, pp.inactiveFlag, pp.activeFlag, 0)
+			pp.ns.SetFieldSub(f.node, pp.capacityField, nil)
 		} else {
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:478]: Shouldn't it be set to 0?

**Review by alisscco — APPROVED** [review]
I really appreciate the work you guys do .
Thank you for changing the future .

File: les/server.go [hunk 10]
```diff
@@ -149,8 +137,12 @@ func NewLesServer(node *node.Node, e ethBackend, config *ethconfig.Config) (*Les
 		srv.maxCapacity = totalRecharge
 	}
 	srv.fcManager.SetCapacityLimits(srv.minCapacity, srv.maxCapacity, srv.minCapacity*2)
-	srv.clientPool = newClientPool(ns, lesDb, srv.minCapacity, defaultConnectedBias, mclock.System{}, srv.dropClient, issync)
-	srv.clientPool.setDefaultFactors(vfs.PriceFactors{TimeFactor: 0, CapacityFactor: 1, RequestFactor: 1}, vfs.PriceFactors{TimeFactor: 0, CapacityFactor: 1, RequestFactor: 1})
+	srv.clientPool = vfs.NewClientPool(lesDb, srv.minCapacity, defaultConnectedBias, mclock.System{}, issync)
+	srv.clientPool.AddMetrics(totalConnectedGauge, clientConnectedMeter, clientDisconnectedMeter,
```
Comment (rjl493456442) [inline les/server.go:141]: Please move these metrics to the clientpool package.

File: les/vflux/server/clientpool.go [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+			if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+				c.Disconnect()
+			}
+		}
+	})
+
+	ns.SubscribeField(btSetup.balanceField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue != nil {
+			ns.SetStateSub(node, ppSetup.inactiveFlag, nodestate.Flags{}, 0)
+			cp.lock.RLock()
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:150]: I think we should put the default price policies into the balance tracker. So that all the newly created balances will be assigned automatically, instead of doing here.

File: les/vflux/server/clientpool.go [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+			}
+		}
+		if newState.Equals(nodestate.Flags{}) {
+			if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+				c.Disconnect()
+			}
+		}
+	})
+
+	ns.SubscribeField(btSetup.balanceField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:147]: Why not moving the "inactive flag tagging" work to the prioritypool? 

Whenever the client field is set, then the balance is created and inactive flag is tagged

File: les/vflux/server/clientpool.go [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+			cp.lock.RLock()
+			newValue.(*nodeBalance).SetPriceFactors(cp.defaultPosFactors, cp.defaultNegFactors)
+			cp.lock.RUnlock()
+		}
+	})
+
+	ns.SubscribeField(ppSetup.capacityField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+			newCap, _ := newValue.(uint64)
+			c.UpdateCapacity(newCap, node == cp.capReqNode)
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:159]: This notification work should be done in the `SetCapacity`

File: les/vflux/server/clientpool.go [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+	capReqNode *enode.Node // node that is requesting capacity change; only used inside NSM operation
+
+	capacityQueryZeroMeter, capacityQueryNonZeroMeter metrics.Meter
+}
+
+// clientPeer represents a peer in the client pool. None of the callbacks should block.
+type clientPeer interface {
+	Node() *enode.Node
+	FreeClientId() string                         // unique id for non-priority clients (typically a prefix of the network address)
+	InactiveTimeout() time.Duration               // disconnection timeout for inactive non-priority peers
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:94]: I would prefer the name `InactiveAllowance`?

File: les/vflux/server/clientpool.go [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+// Unregister removes the peer from the client pool
+func (cp *ClientPool) Unregister(peer clientPeer) {
+	cp.ns.SetField(peer.Node(), clientField, nil)
+}
+
+// SetDefaultFactors sets the default price factors applied to subsequently connected clients
+func (cp *ClientPool) SetDefaultFactors(posFactors, negFactors PriceFactors) {
+	cp.lock.Lock()
+	cp.defaultPosFactors = posFactors
+	cp.defaultNegFactors = negFactors
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:220]: The price policies can be propagated to the balance tracker. So that the newly created balance can be assigned directly.

File: les/vflux/server/prioritypool.go [hunk 16]
```diff
@@ -247,27 +239,42 @@ func (pp *PriorityPool) SetLimits(maxCount, maxCap uint64) {
...
 	}
 }
 
-// SetActiveBias sets the bias applied when trying to activate inactive nodes
-func (pp *PriorityPool) SetActiveBias(bias time.Duration) {
+// setActiveBias sets the bias applied when trying to activate inactive nodes
+func (pp *priorityPool) setActiveBias(bias time.Duration) {
 	pp.lock.Lock()
-	defer pp.lock.Unlock()
+	var updates []capUpdate
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:249]: nitpicks, for this simple function, we can unlock first and then update flags in the end of the function

**Review by alisscco — APPROVED** [review]

File: les/vflux/server/prioritypool.go (hunk 16 above)
Comment (zsfelfoldi) [reply les/vflux/server/prioritypool.go]: done

File: les/vflux/server/clientpool.go (hunk 15 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: done

File: les/vflux/server/clientpool.go (hunk 14 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: done

File: les/vflux/server/clientpool.go (hunk 12 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: Right, it makes no sense anymore to set it separately because now we don't need to set the client field to operate on disconnected node balances.

File: les/vflux/server/clientpool.go (hunk 11 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: done

File: les/server.go (hunk 10 above)
Comment (zsfelfoldi) [reply les/server.go]: done

File: les/vflux/server/balance.go (hunk 8 above)
Comment (zsfelfoldi) [reply les/vflux/server/balance.go]: done

File: les/vflux/server/clientpool.go (hunk 5 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: removed

File: les/vflux/server/clientpool.go (hunk 4 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: done

File: les/vflux/server/prioritypool.go (hunk 9 above)
Comment (zsfelfoldi) [reply les/vflux/server/prioritypool.go:439]: No, the capacity field is never zero, it's `nil` if the node is inactive.

File: les/vflux/server/clientpool.go (hunk 6 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go:233]: I did think a bit about this and this is the best solution I have found so far. The reason I am using a callback here is because the capacity update can be initiated either from the client or server side. If the priority pool reduces the capacity then the server handler needs to be notified in a callback so that it can update its flow control params and send an update message to the client. If the client requests a capacity update and the server handler calls `SetCapacity` then basically the same thing should happen, the same callback will do the update job. The alternative would be to disable calling the callback when it was initiated with `SetCapacity` and leave it to the server handler to call `clientPeer.UpdateCapacity`. That would usually be fine but it is possible that a capacity granted by `SetCapacity` is instantly reduced by the priority pool (despite the bias that usually prevents this). In this case the order of events could be mixed up for the server handler, the callback resulting from the priority pool's reduction could be processed before the update made my the server handler after `SetCapacity` has returned. So I felt like this is the cleaner way.

File: les/vflux/server/clientpool.go (hunk 13 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go:135]: No, see above, the callback is necessary because the capacity can be changed any time (when another client connects or requests more capacity).

File: les/vflux/server/clientpool.go (hunk 7 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go]: I did not really like this part either so I changed it a bit. Now I think it's much nicer, I also added some explanatory comments. So basically it's the `priorityPool` that determines the final capacity (as it should be), `enforceLimits` adjusts the requesting node in finer steps than the rest of the nodes but otherwise it's the same mechanism. The capacity curve estimation is just an optimization layer on top of it, it prevents requesting too high capacity because that could cost a lot of iterations in `enforceLimits`. The estimation itself is very quick and it guarantees that `enforceLimits` will either find the best possible capacity in 5 iterations or fail if there is something wrong with the estimation (this case is detected by the fuzzer).

File: les/vflux/server/prioritypool.go (hunk 2 above)
Comment (zsfelfoldi) [reply les/vflux/server/prioritypool.go]: Actually there is no need to return any kind of priority value any more because the capacity logic was simplified. `setCap` parameter is also not used anymore because there is the capacity curve to efficiently serve pool queries. So I simplified it, now it only returns the final capacity.

File: les/vflux/server/prioritypool.go (hunk 3 above)
Comment (zsfelfoldi) [reply les/vflux/server/prioritypool.go]: Maybe it could be made nicer, I need to think a bit about it.

File: les/vflux/server/clientpool.go (hunk 13 above)
Comment (rjl493456442) [reply les/vflux/server/clientpool.go:135]: I see. Btw I assume that `requested` param will be used in the future? It's not used now. Just want to confirm it's not a mistake.

File: common/prque/sstack.go [hunk 17]
```diff
@@ -94,7 +96,11 @@ func (s *sstack) Len() int {
 // Compares the priority of two elements of the stack (higher is first).
 // Required by sort.Interface.
 func (s *sstack) Less(i, j int) bool {
-	return (s.blocks[i/blockSize][i%blockSize].priority - s.blocks[j/blockSize][j%blockSize].priority) > 0
+	a, b := s.blocks[i/blockSize][i%blockSize].priority, s.blocks[j/blockSize][j%blockSize].priority
+	if s.wrapAround {
```
Comment (rjl493456442) [inline common/prque/sstack.go:100]: What's the difference between the `a>b` and `a-b>0`?

**Review by alisscco — APPROVED** [review]

File: les/vflux/server/clientpool.go [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+				// Note: if capacity is immediately available then priorityPool will set the active
+				// flag simultaneously with removing the inactive flag and therefore this will not
+				// initiate disconnection
+				ns.SetStateSub(node, nodestate.Flags{}, setup.inactiveFlag, 0)
+			}
+		}
+		if oldState.Equals(setup.inactiveFlag) && newState.Equals(setup.inactiveFlag.Or(setup.priorityFlag)) {
+			ns.SetStateSub(node, setup.inactiveFlag, nodestate.Flags{}, 0) // priority gained; remove timeout
+		}
+		if newState.Equals(setup.activeFlag) {
```
Comment (rjl493456442) [inline les/vflux/server/clientpool.go:118]: I thought about it. All the client will be assigned an initial capacity, it's minimal cap. This check is just useless?

**Review by alisscco — APPROVED** [review]

**Review by alisscco — APPROVED** [review]

**Comment by holiman** [comment]
@alisscco please stop adding random approvals to PRs

File: les/vflux/server/clientpool.go (hunk 18 above)
Comment (zsfelfoldi) [reply les/vflux/server/clientpool.go:118]: This check is not about newly activated clients. It reverts higher capacity priority clients to `minCap` if they lose their priority flag. `newState.Equals(setup.activeFlag)` means active, but has no priority flag (no positive balance).

File: common/prque/sstack.go (hunk 17 above)
Comment (zsfelfoldi) [reply common/prque/sstack.go:100]: The way they handle numeric overflows. For example, `MaxInt64 > MinInt64` but `MaxInt64-MinInt64 < 0`. If we use the latter type of comparison then the individual priority values are allowed to overflow and "wrap around" the 64 bit range many times as long as the biggest difference between two priorities in the queue at any moment can be expressed in the `int64` range. `flowcontrol.ClientManager` is a good example for this mode where there is a cumulative "buffer recharge integrator" that is continuously being increased and does overflow. `rcQueue` item priorities are all calculated relative to this value and the difference between any two item priorities fits safely into `int64`. The first (free clients only) client pool design also did take advantage of this feature but the current priority pool scheme uses absolute values.
This "wrap around" feature requires care to use and until now all `prque.Prque`-s operated in this mode. Now it's only used if the queue is explicitly created with the `NewWrapAround` constructor.

Comment (zsfelfoldi) [reply common/prque/sstack.go:100]: This PR reverts to the non-tricky version in most of the cases and does not change the behavior of the client manager. I will change client manager's priority calculation and get rid of thip wrap-around trick altogether in a next PR.

Comment (rjl493456442) [reply common/prque/sstack.go:100]: Thanks!

File: les/vflux/server/prioritypool.go (hunk 3 above)
Comment (zsfelfoldi) [reply les/vflux/server/prioritypool.go]: Now I think that the temporary state does make sense because in some cases I do call `enforceLimits` multiple times and decide the next action based on the temporary state resulting from the previous iteration. It was probably hard to follow though how the temporary state works so I added a commit that does some supposedly identical changes that hopefully makes it cleaner. Now the `capacity` field is not changed until the temp state is committed and there is a `tempCapacity` so it's always clear what we are operating on.

File: les/vflux/server/prioritypool.go [hunk 19]
```diff
@@ -320,36 +324,45 @@ func (pp *priorityPool) disconnectedNode(c *ppNodeInfo) {
...
+func (pp *priorityPool) setTempState(c *ppNodeInfo) {
+	if c.tempState {
 		return
 	}
-	c.changed = true
-	c.origCap = c.capacity
-	pp.changed = append(pp.changed, c)
+	c.tempState = true
+	if c.tempCapacity != c.capacity { // should never happen
+		log.Crit("tempCapacity != capacity when entering tempState")
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:343]: Can we lower this log level to Error? I am afraid that the LES part has some issues and will just shutdown the entire Geth node.

File: les/vflux/server/prioritypool.go [hunk 20]
```diff
@@ -320,36 +324,45 @@ func (pp *priorityPool) disconnectedNode(c *ppNodeInfo) {
...
-// be called after markForChange and before finalizeChanges.
-func (pp *priorityPool) setCapacity(n *ppNodeInfo, cap uint64) {
-	pp.activeCap += cap - n.capacity
-	if n.capacity == 0 {
+// setTempCapacity changes the capacity of a node in the temporary state and adjusts
+// activeCap and activeCount accordingly. Since this change is performed in the temporary
+// state it should be called after setTempState and before finalizeChanges.
+func (pp *priorityPool) setTempCapacity(n *ppNodeInfo, cap uint64) {
+	if !n.tempState { // should never happen
+		log.Crit("Node is not in temporary state")
```
Comment (rjl493456442) [inline les/vflux/server/prioritypool.go:355]: Can we lower this log level to Error? I am afraid that the LES part has some issues and will just shutdown the entire Geth node.

**Review by rjl493456442 — APPROVED** [review]
LGTM

**Comment by zsfelfoldi** [comment]
@fjl PTAL at the `GetState` function in `NodeStateMachine`. It works similarly to the already existing `GetField`.

**Review by fjl — APPROVED** [review]

**Review by slumericanB — COMMENTED** [review]
(y)
